man/censuslookup.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "CENSUSLOOKUP" "1" "10 June 2026" "" "The Regina Handbook"

.SH NAME
censuslookup \- Search for triangulations or link diagrams in Regina's census databases
.SH SYNOPSIS

\fBcensuslookup\fR \fB\fIisosig\fB\fR\fI ...\fR


\fBcensuslookup\fR { \fB-v, --version\fR | \fB-?, --help\fR }

.SH "DESCRIPTION"
.PP
This utility searches for the given 3-manifold triangulations and/or link
diagrams in the various census databases that are shipped with Regina.
.PP
You should present trianguations using their
\fIisomorphism signatures\fR (short text strings that
identify triangulations uniquely up to relabelling of their tetrahedra
and/or their vertices).
You should present link diagrams using their
\fIknot/link signatures\fR (short text strings that
identify link diagrams uniquely up to relabelling, reflection and/or
rotation of the entire diagram, and/or reversal of individual link
components).
.PP
You can use first-generation signatures (as used by
Regina\~<=\~7.x) or second-generation signatures (as
introduced in Regina\~8.0); either will be understood.
You can also include a mix of both triangulations and links in the
same call to \fBcensuslookup\fR\&.
.PP
From the graphical user interface, you can view isomorphism signatures
by switching to the Composition tab in the
triangulation viewer, and you can view knot/link signatures by switching
to the Codes tab in the link viewer.
From Python you can fetch the second-generation signature of a
triangulation or link \fBx\fR by calling
\fBx.neoSig()\fR\&.
If you need a first-generation signature for compatibility with older
versions of Regina, instead call \fBx.isoSig()\fR
for triangulations, or \fBx.knotSig()\fR for links.
.PP
For a full and precise specification of first-generation isomorphism
signatures of 3-manifold triangulations, see
\fBSimplification paths in the Pachner graphs of closed
orientable 3-manifold triangulations\fR, Burton, 2011,
arXiv:1110.6080.
.PP
All database matches (if any) will be written to standard output.
.SH "OPTIONS"
.TP
\fB-v, --version\fR
Show which version of Regina is being used, and exit
immediately.
.TP
\fB-?, --help\fR
Display brief usage information, and exit immediately.
.SH "ENVIRONMENT VARIABLES"
.PP
The following environment variables influence the behaviour of
this program.
.TP
\fB\fIREGINA_HOME\fB\fR
The directory beneath which Regina's data files are installed.
This utility will look for the census databases in the subdirectory
\fI$REGINA_HOME/data/census/\fR\&.

Normally you should not need to set this option yourself.
This utility should know how to find its census databases in standard
situations, which include
fixed filesystem installations (e.g., \fBGNU/Linux\fR and \fBWindows\fR),
relocatable app bundles (e.g., \fBmacOS\fR),
and running directly from the source tree.
.SH "EXAMPLE"
.PP
The following example uses second-generation isomorphism signatures:

.nf
    example$ \fBcensuslookup fFuC8hYTOTa dl1hXPzb cV6cqb\fR
    fFuC8hYTOTa: 1 hit
        SFS [S2: (2,1) (3,1) (5,-4)] : #1 -- Closed census (orientable)

    dl1hXPzb: 0 hits

    cV6cqb: 2 hits
        m004 : #1 -- Cusped hyperbolic census (orientable)
        L104001 -- Hyperbolic knot and link complements

    example$
.fi
.PP
The next example describes the same triangulations as before, but this
time using first-generation isomorphism signatures:

.nf
    example$ \fBcensuslookup fvPQcdecedekrsnrs dLQbcccajqs cPcbbbiht\fR
    fvPQcdecedekrsnrs: 1 hit
        SFS [S2: (2,1) (3,1) (5,-4)] : #1 -- Closed census (orientable)

    dLQbcccajqs: 0 hits

    cPcbbbiht: 2 hits
        m004 : #1 -- Cusped hyperbolic census (orientable)
        L104001 -- Hyperbolic knot and link complements

    example$
.fi
.PP
This example uses knot/link signatures:

.nf
    example$ \fBcensuslookup eputWe\fR
    eputWe: 3 hits
        4ah_1 -- Prime classical knots
        v4_1 -- Prime virtual knots
        4.108 -- Green's virtual knot tables

    example$
.fi
.SH "MACOS USERS"
.PP
If you downloaded a drag-and-drop app bundle, this utility is
shipped inside it.  If you dragged Regina to the main
Applications folder, you can run it as
/Applications/Regina.app/Contents/MacOS/censuslookup\&.
.SH "WINDOWS USERS"
.PP
The command-line utilities are installed beneath the
\fIProgram\~Files\fR directory; on some
machines this directory is called
\fIProgram\~Files\~(x86)\fR\&.
You can start this utility by running
c:\\Program\~Files\\Regina\\Regina\~8.0\\bin\\censuslookup.exe\&.
.SH "SEE ALSO"
.PP
regina-gui\&.
trisetcmp,
.SH "AUTHOR"
.PP
This utility was written by Benjamin Burton
<bab@maths.uq.edu.au>\&.
Many people have been involved in the development
of Regina; see the users' handbook for a full list of credits.


utils/censuslookup.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Look up an isomorphism signature in Regina's census databases         *
 *                                                                        *
 *  Copyright (c) 2014-2026, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program. If not, see <https://www.gnu.org/licenses/>. *
 *                                                                        *
 **************************************************************************/

#include <cstring>
#include <iostream>
#include "census/census.h"
#include "file/globaldirs.h"
#include "utilities/exception.h"

void usage(const char* progName, const std::string& error = std::string()) {
    if (! error.empty())
        std::cerr << error << "\n\n";

    std::cerr << "Usage:" << std::endl;
    std::cerr << "    " << progName << " <sig> ..." << std::endl;
    std::cerr << "    " << progName
        << " [ -v, --version | -?, --help ]" << std::endl;
    std::cerr << std::endl;
    std::cerr << "    -v, --version : Show which version of Regina "
        "is being used" << std::endl;
    std::cerr << "    -?, --help    : Display this help" << std::endl;
    exit(1);
}

int main(int argc, char* argv[]) {
    // Parse the command line.
    if (argc < 2)
        usage(argv[0], "Please specify one or more isomorphism signatures "
            "or knot/link signatures.");

    // Check for standard arguments:
    for (int i = 1; i < argc; ++i) {
        if (strcmp(argv[i], "-?") == 0 || strcmp(argv[i], "--help") == 0)
            usage(argv[0]);
        if (strcmp(argv[i], "-v") == 0 || strcmp(argv[i], "--version") == 0) {
            if (argc != 2)
                usage(argv[0],
                    "Option --version cannot be used with "
                        "any other arguments.");
            std::cout << PACKAGE_BUILD_STRING << std::endl;
            exit(0);
        }
    }

    // Locate the census data files.
    regina::GlobalDirs::deduceDirs(argv[0]);

    // Search for each signature.
    try {
        for (int i = 1; i < argc; ++i) {
            auto hits = regina::Census::lookup(argv[i]);

            size_t n = hits.size();
            std::cout << argv[i] << ": " << n
                << (n == 1 ? " hit" : " hits") << std::endl;

            for (const auto& hit : hits)
                std::cout << "    " << hit.name() << " -- "
                    << hit.db().desc() << std::endl;

            std::cout << std::endl;
        }
    } catch (const regina::FileError&) {
        std::cerr << "ERROR: An error occurred whilst accessing Regina's "
            "census databases.\n\n";
        std::cerr << "Please check that Regina has been installed correctly.\n";
        return 1;
    }

    return 0;
}
